scripts/validate_config.py

```diff
@@ -0,0 +1,136 @@
+
+    '''
+    Function maintains the arguments accepted by this script. Please see
+    Python's argparse documentation for more information about settings of each
+    argument.
+    '''
+
+    parser = argparse.ArgumentParser(
+       description='Validate config with user-defined settings.'
+    )
+
+    group = parser.add_mutually_exclusive_group()
+
+    parser.add_argument(
+        '-s', '--validation_schema',
+        help='Path to a validation schema.',
+        required=True,
+        type=cli_helpers.path_if_file_exists,
+    )
+
+    parser.add_argument(
+        '-c', '--config_file',
+        help='Path to configuration file. Accepts YAML, bash/ini or namelist',
+        required=True,
+        type=cli_helpers.path_if_file_exists,
+    )
+
+    parser.add_argument(
+        '--config_file_type',
+        help='If used, will convert provided config file to given file type.\
+            Accepts YAML, bash/ini or namelist',
+        choices=["YAML", "INI", "F90"],
+    )
+
+    group.add_argument(
+        '-v', '--verbose',
+        action='store_true',
+        help='If provided, print all logging messages.',
+        )
+    group.add_argument(
+        '-q', '--quiet',
+        action='store_true',
+        help='If provided, print no logging messages',
+        )
+    parser.add_argument(
+        '-l', '--log_file',
+        help='Optional path to a specified log file',
+        default=os.path.join(os.path.dirname(__file__), "set_config.log")
+        )
+
+    return parser.parse_args(argv)
+
+
+def validate_config(argv, log=None):
+    '''Main section for validating config file'''
+
+    user_args = parse_args(argv)
+
+    if log is None:
+        name = f"{inspect.stack()[0][3]}"
+        log = cli_helpers.setup_logging(user_args, log_name=name)
+
+    # Get the config file to be validated and dereference jinja templates
+    config_class = getattr(config, "YAMLConfig")
+    config_obj = config_class(user_args.config_file, log_name=log.name)
+    config_obj.dereference_all()
+
+    # Load the json validation schema
+    with open(user_args.validation_schema, 'r',
+              encoding="utf-8") as schema_file:
+        schema = json.load(schema_file)
+
+    schema_error = 0
+
+    # Validate the config file against the schema file
+    validator = jsonschema.Draft7Validator(schema)
+
+    # Print out each schema error
+    errors = validator.iter_errors(config_obj.data)
+
+    for error in errors:
+        schema_error += 1
+        logging.error(error)
+        logging.error('------')
+
+    # Create a list of fields that could contain a file or path
+    path_list = []
+    for field in schema["properties"]:
+        for value in schema["properties"][field]["properties"]:
+            if "format" in schema["properties"][field]["properties"][value]:
+                path_list.append(value)
+
+    # Check for existence of those files or paths
+    for field in config_obj.data:
+        for value in config_obj.data[field]:
```

Comment:
These two lines could clean up to:
```
for field, value in config_obj.data.items():
   ...
```

Reply:
The "field, value" code returns the key and the value, which is another dictionary. The second level dictionary also needs to be iterated through. In the original code, the first for loop is iterating through the YAML sections, and the second for loop is iterating through each field in a section looking for paths and files.

Reply:
I know what @christinaholtNOAA meant, but see that @venitahagerty is right. Since the outermost keys are not actually needed (except to get their values), an option might be
```
for section in config_obj.data.values():
    for key, value in section.items():
        ...
```
e.g.
```
>>> s = """
... fruit-colors:
...   apple: red
...   banana: yellow
...   pear: green
... pet-names:
...   dog: spot
...   parrot: polly
... """
>>> d = yaml.load(s, Loader=yaml.Loader)
>>> for section in d.values():
...     for key, value in section.items():
...         if value == "yellow":
...             print(f"{key} is yellow")
... 
banana is yellow
```
I'm not sure if `section`, `key`, and `value` are good names here, or if I've got the structure of the input data right, but this potentially eliminates some repetitive `[]` `dict` lookups.

This brings up a question for me though: Does our input data currently have only two levels of nesting, and is that something we think we can maintain in the future?

Reply:
Christina and I discussed that, and she said there would only be 2 levels of nesting.

Reply:
Ahh, yes. For these machine files, we do have a limitation that this is as deep as it can go. I think we will want to delve into recursion at some point since that this limited depth is only a requirement for SR , but it is likely out of scope of this ticket. We have a lot of additional schema validator work on the backlog already.


Also, I totally missed the boat on the nested loops above. Thanks to both of you for the clarification.

Reply:
I do think using `.values()` and `.items()` is cleaner and that we should update this code later to use `section` and `value` in place of `config_obj.data[field]` and (2 copies of) `config_obj.data[field][value]`. Later...